### ISSUE TITLE
Separate storage backend for remote media

### ISSUE THREAD

**Issue by lullis** [mastodon/mastodon#28554]
### Pitch

Having a replica of every image on every instance is quite a bit wasteful. I understand the design goals around it to avoid clients querying a remote instance directly which might not be trusted, but I believe we could *at the very least* have a way to split between media from the local instance (which we can not / should not delete unless the user wants to) and remote media which can be refetched.

### Motivation

With separate media storages, admins could come up with different strategies for the different types of media:

 - S3 buckets with unlimited capacity for local data, S3 buckets capped at  X GB for remote media
 - S3 storage for local data, local disk volumes from cloud providers for remote media
 - mirrored RAID 1 volumes for local data, JBOD array for remote data.
 - object storage for local data, content addressable systems (like IPFS or magnet links) for remote data.

**Comment by KuJoe** [comment]
I agree with this, it would be nice to keep all of the critical, local media stored locally with the servers on more expensive storage while keeping the cached/remote data hosted on cheaper object storage like S3 which is also more prone to outages/connection issues (99% of my issues have been a result of object storage, all of which have been out of my control due to my budget).

My current remote storage is nearing 2TB for 30 day's worth of cache but less than 120GB of that is local data that needs to be accessible at all times for my users (and keeping it local will greatly improve performance compared to the 80ms of latency to my current object storage).

This suggestion is an overall stability, performance, and financial improvement.

**Comment by smiba** [comment]
Would be really useful, we're considering the use of a S3 proxy setup (using something like nginx-s3-gateway) so we can have the cache directory on local S3 storage hosted with MinIO, while storing user uploads with higher redundancy remotely.

Biggest issue is that to make this work you have to recalculate the s3 signatures, which is a PITA. Would be much easier if mastodon itself could just connect to the right endpoint.

**Comment by smiba** [comment]
Update to this:

Ended up making a nginx reverse proxy that can route S3 endpoints based on the file location using OpenResty (nginx + lua) and the following repo: https://github.com/smiba/lua-resty-aws

Currently testing just a little more in our dev environment but so far as so good, it allows us to pick a different storage endpoint for cache files.

OpenResty config file:
```
worker_processes  1;

events {
    worker_connections  1024;
}

http {
    sendfile        on;

    map $request_uri $request_uri_no_parameters {
      "~^(?P<path>.*?)(\?.*)*$"  $path;
    }

    server {
      listen 127.0.0.1:8080;

      client_max_body_size 1G;
      client_body_buffer_size 1024M;
      client_body_in_single_buffer on;

      set $aws_uri $request_uri_no_parameters;

      lua_need_request_body on;

      location /stor-test/cache/ {
        set $aws_host storage.example.com;
        set $aws_access_key <your AWS access key for this endpoint here>;
        set $aws_secret_key <your AWS secret key for this endpoint here>;
        set $aws_service s3;
        set $aws_region nl-ams;

        access_by_lua "local aws = require 'resty.aws'; aws.s3_set_headers(ngx.var.aws_access_key, ngx.var.aws_secret_key, ngx.var.aws_host, ngx.var.aws_uri, ngx.var.aws_region, ngx.var.aws_service)";

        proxy_pass https://storage.example.com;
      }

      location /stor-test/ {
        set $aws_host s3.nl-ams.scw.cloud;
        set $aws_access_key <your AWS access key for this endpoint here>;
        set $aws_secret_key <your AWS secret key for this endpoint here>;
        set $aws_service s3;
        set $aws_region nl-ams;

        access_by_lua "local aws = require 'resty.aws'; aws.s3_set_headers(ngx.var.aws_access_key, ngx.var.aws_secret_key, ngx.var.aws_host, ngx.var.aws_uri, ngx.var.aws_region, ngx.var.aws_service)";
        proxy_pass https://s3.nl-ams.scw.cloud;
      }

      location / {
        return 404; #Not our bucket?
      }
    }
}
```

Only issue is that commands that list the entire bucket's content (such as tootctl media remove-orphans) do no longer work, as it does not understand the content is split between endpoints. AFAIK this is the only command that uses a directory listing command and this is not used further inside regular mastodon workings. (Please let me know if it's not though! I'd like to be corrected before I push this to production and figure it out the hard way!)

If anyone else is interested in this setup, I can do a proper writeup on how to set it up and how it works in general.

**Comment by smiba** [comment]
Published a write-up on how we're now using two S3 endpoints: https://medium.com/@PurpleDug/woof-tech-routing-s3-endpoints-and-doing-a-live-migration-on-a-production-environment-as-a-treat-8fa2bbd6784e

Would still love to see official support within mastodon itself for this, but for now this is working well for us :)
